src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -0,0 +1,370 @@
+                {
+                    return false;
+                }
+
+                return await project.HasSuccessfullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            private static bool IsCanceled(Exception ex, CancellationToken cancellationToken)
+            {
+                return (ex as OperationCanceledException)?.CancellationToken == cancellationToken;
+            }
+
+            private void OnAnalyzerException(DiagnosticAnalyzer analyzer, ProjectId projectId, Compilation compilationOpt, Exception ex)
+            {
+                var exceptionDiagnostic = AnalyzerHelper.CreateAnalyzerExceptionDiagnostic(analyzer, ex);
+
+                if (compilationOpt != null)
+                {
+                    exceptionDiagnostic = CompilationWithAnalyzers.GetEffectiveDiagnostics(ImmutableArray.Create(exceptionDiagnostic), compilationOpt).SingleOrDefault();
+                }
+
+                var onAnalyzerException = _owner.GetOnAnalyzerException(projectId);
+                onAnalyzerException(ex, analyzer, exceptionDiagnostic);
+            }
+
+            private IEnumerable<DiagnosticData> ConvertToLocalDiagnosticsWithoutCompilation(Document targetDocument, IEnumerable<Diagnostic> diagnostics, TextSpan? span = null)
+            {
+                var project = targetDocument.Project;
+                Contract.ThrowIfTrue(project.SupportsCompilation);
+
+                foreach (var diagnostic in diagnostics)
+                {
+                    var location = diagnostic.Location;
+                    if (location.Kind != LocationKind.ExternalFile)
+                    {
+                        continue;
+                    }
+
+                    var lineSpan = location.GetLineSpan();
+
+                    var documentIds = project.Solution.GetDocumentIdsWithFilePath(lineSpan.Path);
+                    if (documentIds.IsEmpty || documentIds.All(id => id != targetDocument.Id))
+                    {
+                        continue;
+                    }
+
+                    yield return DiagnosticData.Create(targetDocument, diagnostic);
+                }
+            }
+
+            private IEnumerable<DiagnosticData> ConvertToLocalDiagnosticsWithCompilation(Document targetDocument, IEnumerable<Diagnostic> diagnostics, TextSpan? span = null)
+            {
+                var project = targetDocument.Project;
+                Contract.ThrowIfFalse(project.SupportsCompilation);
+
+                foreach (var diagnostic in diagnostics)
+                {
+                    var document = project.GetDocument(diagnostic.Location.SourceTree);
+                    if (document == null || document != targetDocument)
+                    {
+                        continue;
+                    }
+
+                    if (span.HasValue && !span.Value.Contains(diagnostic.Location.SourceSpan))
```

Comment:
v1 checks for Intersects: http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs,218, not sure if that matters